content/posts/deta_fastapi_jwt_auth_part_2.md

```diff
@@ -0,0 +1,273 @@
+    if users_db.get(user_details.username) != None:
+        return 'Account already exists'
+    try:
+        hashed_password = auth_handler.encode_password(user_details.password)
+        user = {'key': user_details.username, 'password': hashed_password}
+        return users_db.put(user)
+    except:
+        error_msg = 'Failed to signup user'
+        return error_msg
+```
+
+In this function we are checking if a user with the username already exists in our `users_db`. If so, we can simply return a message indicating that the account already exists. If the user doesn't already exists we can hash the password using the `encode_password` function from `auth.py` and store the user in `users_db`. In case of any errors while adding the user to the base, we can return a failure message. 
+
+`/login` endpoint is pretty simple. This also takes in the argument `user_details` , which has the username and password. 
+
+```python
+@app.post('/login')
+def login(user_details: AuthModal):
+    user = users_db.get(user_details.username)
+    if (user is None):
+        return HTTPException(status_code=401, detail='Invalid username')
+    if (not auth_handler.verify_password(user_details.password, user['password'])):
+        return HTTPException(status_code=401, detail='Invalid password')
+    
+    token = auth_handler.encode_token(user['key'])
+    return {'token': token}
+```
+
+If the account with the username doesn't exist, or if the hashed password in the `users_db` doesn't match the input password we can simply raise an `HTTPException`. Otherwise, we can return the encoded JWT token using `encode_token`.
+
+```python
+@app.post('/secret')
+def secret_data(credentials: HTTPAuthorizationCredentials = Security(security)):
+    token = credentials.credentials
+    if(auth_handler.decode_token(token)):
+        return 'Top Secret data only authorized users can access this info'
+
+@app.get('/notsecret')
+def not_secret_data():
+    return 'Not secret data'
+```
+
+The `/secret` endpoint only returns the "Secret Data" if the token argument is valid. However, if the token is invalid or an expired token, then `decode_token` raises a `HTTPException`. The token is usually passed in the request header as `Authorization: Bearer <token>`. Therefore, to get the token we can wrap the input `credentials` around `HTTPAuthorizationCredentials` tag. Now we can access the token from the request header in `credentials.credentials`. 
+
+The `/not_secret` endpoint is an example of an unprotected endpoint, which doesn't require any authentication.
+
+```python
+@app.get('/refresh_token')
+def refresh_token(credentials: HTTPAuthorizationCredentials = Security(security)):
+    expired_token = credentials.credentials
+    return auth_handler.refresh_token(expired_token)
+```
+
+`/refresh_token` endpoint is also pretty simple, it receives an expired token which is then passed onto the the `refresh_token` function from auth logic to get the new token. 
```

Comment:
Getting a refresh token from the already expired token is bad. 

The refresh token should be:
- separate and should be provided when you login initially
- should have longer expiry time than the jwt token
- should only have access to get new jwt tokens

Reply:
Sorry I should have looked at it earlier, I see blog part I already has been published with this refresh workflow, I would recommend either getting rid of the refresh workflow (with a note informing that in production it's recommended to have a refresh token) or fixing this refresh workflow.